### ISSUE TITLE
[Bug] SM120 sparse-MLA prefill (DSv4 dual-cache) returns corrupted output for >64 query tokens; masking lanes beyond topk_length to -1 fixes it

### ISSUE THREAD

**Issue by caiovicentino** [flashinfer-ai/flashinfer#5095]
## Summary

On SM120 (4x RTX PRO 6000 Blackwell, no NVLink), the sparse-MLA **prefill** path (`_sparse_mla_sm120_paged_attention` → `sparse_mla_sm120_prefill.cu`, dual-cache DSv4 form) returns corrupted attention output whenever the query length exceeds 64 tokens, i.e. as soon as SGLang routes past the standalone decode kernels. Decode (`num_tokens <= 64`) is correct. The corruption disappears when the caller masks every `indices` / `extra_indices` lane at position `>= topk_length[i]` to `-1` before the call, which matches the "lanes beyond `topk_len` are not bounded" fix described in #5075.

Reporting because there is no open issue for the *silent wrong output* form of this bug (as opposed to hangs), and because it is easy to miss: short prompts are fine.

## Environment

- flashinfer-python **0.6.18** (as pinned by the `lmsysorg/sglang:dev-dsv41` image, 2026-09-10)
- SGLang dev branch for DeepSeek-V4.1 (`dsv4.1` / PR sgl-project/sglang#38798), plus the SM120 page-split shim from 0xSero/deepseek-v4.1-flash-4x-rtx-pro-6000 (`flash_mla_sm120.py`, which splits the ratio-1/2 extra caches from 256/128-token pages to 64-token pages before calling FlashInfer)
- Model: `deepseek-ai/DeepSeek-V4.1-Flash`, `--attention-backend dsv4`, TP4, fp8 KV, DSpark spec decoding (bug reproduces with spec decoding disabled as well)
- CUDA driver 13000, torch 2.13+cu130, 4x RTX PRO 6000 (SM120), Linux

## Symptom

Chat completions at `temperature=0`:

| prompt tokens | output |
|---|---|
| 18 … 62 | correct, deterministic |
| 65 | garbage (`](). The batch size is 32, the learning rate is 1e-4 …`) |
| 69 | garbage (` */, and why the role of the KV cache …`) |
| 122 (English) | unrelated text (`(Redirected from Wikipedia:Reference desk …`) |
| 160 (Portuguese) | repeated CJK fragment in a loop |

Long *generation* from a short prompt (300+ tokens decoded, so decode steps run well past position 64) is always correct, which isolates the prefill kernel. Under some configs (EP1 + marlin MoE) the garbage was **non-deterministic** across two identical `temperature=0` requests, consistent with reading uninitialized/stale memory rather than a deterministic layout mistake.

The threshold is exactly the `_DECODE_MAX_TOKENS = 64` boundary in `flashinfer/mla/_sparse_mla_sm120.py`.

## What rules things out

- Same corruption with EP1 and EP4, with `flashinfer_mxfp4` and `marlin` MoE runners, with and without speculative decoding, and with the Engram tables served from RAM or NVMe.
- `SGLANG_SM120_FLASHMLA_BACKEND=triton` (SGLang's Triton sparse-decode kernel used for every call) → fully correct output at every prompt length. Also correct with a hybrid: FlashInfer for `num_tokens <= 64`, Triton for `> 64`.
- **Correct output is restored with the FlashInfer prefill path** by masking the padding lanes before the call (SGLang side, in `_flash_mla_sm120_prefill`):

```python
if topk_length is not None:
    ar = torch.arange(idx.shape[-1], device=idx.device)
    idx = torch.where(ar[None, :] < topk_length.view(-1, 1), idx, torch.full_like(idx, -1))
if extra_idx is not None and extra_topk_length is not None:
    ar = torch.arange(extra_idx.shape[-1], device=extra_idx.device)
    extra_idx = torch.where(ar[None, :] < extra_topk_length.view(-1, 1), extra_idx, torch.full_like(extra_idx, -1))
```

With that mask, a 16-point bisection (18 → 72 prompt tokens), long-context requests up to 29k tokens and a 5.8k-token generation are all correct, and prefill throughput is ~6.4k tok/s vs ~2.5k tok/s on the Triton fallback.

So the caller was passing lanes beyond `topk_length` that hold stale (not `-1`) slot IDs, and the prefill orchestrator dereferences them. The decode kernels (or SGLang's torch masking on the decode path) do not, hence "decode fine, prefill broken".

## Expected

Either the prefill launchers honor `topk_length` / `extra_topk_length` (bound the lanes, as #5075 proposes), or the Python wrapper documents that lanes past `topk_length` must be `-1` and asserts / masks accordingly, so the failure is loud instead of silent.

## Minimal repro (server-level)

Send the same chat request with the prompt padded one sentence at a time and watch for the output to flip at 63–65 prompt tokens. I can share the 30-line bisection script and the exact prompts if useful, and I am happy to test a patched wheel on this hardware.

**Comment by bkryu** [comment]
Thanks @caiovicentino. As you mentioned in the issue description, we expect #5075 to fix the current issue
